File: libs/server/persistent/src/persistent/prisma/workspace-inbox.ts
import { randomUUID } from 'node:crypto';
import {
  assertInboxVersion,
  DomainError,
  InboxItem,
  InboxRevision,
  Ref,
  type CapturedInboxItem,
  type InboxContentType,
  type InboxItemStatus,
  type InboxListQuery,
  type InboxSourceInput,
  type JsonValue,
  type WorkspaceInbox,
} from '@evidence/server-domain';
import { Prisma } from '@prisma/client';
import { EntityList } from '../database';
import { hashInboxSource, type NormalizedInboxSource } from '../inbox-content';
import type { PrismaStore } from './types';
import { inputJson, isUniqueConflict } from './utils';

const ITEM_INCLUDE = {
  latestRevision: true,
  _count: { select: { revisions: true } },
} satisfies Prisma.InboxItemInclude;

type InboxItemRow = Prisma.InboxItemGetPayload<{
  include: typeof ITEM_INCLUDE;
}>;
type InboxRevisionRow = Prisma.InboxRevisionGetPayload<Record<string, never>>;

export class PrismaWorkspaceInbox
  extends EntityList<InboxItem>
  implements WorkspaceInbox
{
  constructor(
    private readonly store: PrismaStore,
    private readonly workspaceId: string,
  ) {
    super();
  }

  protected override async findEntities(
    from: number,
    to: number,
  ): Promise<InboxItem[]> {
    const rows = await this.store.inboxItem.findMany({
      where: { workspaceId: this.workspaceId },
      include: ITEM_INCLUDE,
      orderBy: { updatedAt: 'desc' },
      skip: from,
      take: Math.max(to - from, 0),
    });
    return rows.map(assembleInboxItem);
  }

  protected override async findEntity(id: string): Promise<InboxItem | null> {
    const row = await this.store.inboxItem.findFirst({
      where: { id, workspaceId: this.workspaceId },
      include: ITEM_INCLUDE,
    });
    return row ? assembleInboxItem(row) : null;
  }

  override async size(): Promise<number> {
    return this.store.inboxItem.count({
      where: { workspaceId: this.workspaceId },
    });
  }

  async list(query: InboxListQuery): Promise<[InboxItem[], number]> {
    validatePage(query.page, query.pageSize);
    const where = listWhere(this.workspaceId, query);
    const [rows, total] = await Promise.all([
      this.store.inboxItem.findMany({
        where,
        include: ITEM_INCLUDE,
        orderBy: { updatedAt: 'desc' },
        skip: (query.page - 1) * query.pageSize,
        take: query.pageSize,
      }),
      this.store.inboxItem.count({ where }),
    ]);
    return [rows.map(assembleInboxItem), total];
  }

  async capture(sourceInput: InboxSourceInput): Promise<CapturedInboxItem> {
    const { source, contentSha256 } = hashInboxSource(sourceInput);
    const itemId = randomUUID();
    const revisionId = randomUUID();
    const timestamp = new Date();

    try {
      await this.transaction(async (store) => {
        await store.inboxItem.create({
          data: {
            id: itemId,
            workspaceId: this.workspaceId,
            sourceKind: source.sourceKind,
            externalKey: source.externalKey,
            title: source.title,
            status: 'active',
            latestRevisionId: null,
            version: 1,
            createdAt: timestamp,
            updatedAt: timestamp,
          },
        });
        await store.inboxRevision.create({
          data: revisionData(
            revisionId,
            itemId,
            1,
            source,
            contentSha256,
            timestamp,
          ),
        });
        await store.inboxItem.update({
          where: { id: itemId },
          data: { latestRevisionId: revisionId },
        });
      });
    } catch (error) {
      if (isUniqueConflict(error)) {
        throw DomainError.conflict(
          `Inbox source ${source.sourceKind}/${source.externalKey} already exists`,
        );
      }
      throw error;
    }

    return this.captureResult(itemId, revisionId, true);
  }

  async appendRevision(
    itemId: string,
    sourceInput: InboxSourceInput,
    expectedLatestRevisionSha256?: string,
  ): Promise<CapturedInboxItem> {
    const { source, contentSha256 } = hashInboxSource(sourceInput);
    let revisionId = '';
    let revisionCreated = false;

    try {
      await this.transaction(async (store) => {
        const current = await requireItem(store, this.workspaceId, itemId);
        const latestRevision = requireLatestRevision(current);
        assertSameSource(current, source);
        if (
          expectedLatestRevisionSha256 !== undefined &&
          latestRevision.contentSha256 !== expectedLatestRevisionSha256
        ) {
          throw DomainError.conflict(
            `Inbox item ${itemId} latest revision has changed`,
          );
        }
        if (latestRevision.contentSha256 === contentSha256) {
          revisionId = latestRevision.id;
          return;
        }

        revisionId = randomUUID();
        const timestamp = new Date();
        await store.inboxRevision.create({
          data: revisionData(
            revisionId,
            itemId,
            current._count.revisions + 1,
            source,
            contentSha256,
            timestamp,
          ),
        });
        const updated = await store.inboxItem.updateMany({
          where: {
            id: itemId,
            workspaceId: this.workspaceId,
            version: current.version,
          },
          data: {
            title: source.title,
            latestRevisionId: revisionId,
            version: { increment: 1 },
            updatedAt: timestamp,
          },
        });
        if (updated.count !== 1) {
          throw DomainError.conflict(`Inbox item ${itemId} has changed`);
        }
        revisionCreated = true;
      });
    } catch (error) {
      if (isUniqueConflict(error)) {
        throw DomainError.conflict(`Inbox item ${itemId} has changed`);
      }
      throw error;
    }

    return this.captureResult(itemId, revisionId, revisionCreated);
  }

  async changeStatus(
    itemId: string,
    status: InboxItemStatus,
    expectedVersion: number,
  ): Promise<InboxItem> {
    assertInboxVersion(expectedVersion);
    const current = await this.findByIdentity(itemId);
    if (!current) {
      throw DomainError.notFound(`Inbox item ${itemId} not found`);
    }
    const updated = await this.store.inboxItem.updateMany({
      where: {
        id: itemId,
        workspaceId: this.workspaceId,
        version: expectedVersion,
      },
      data: {
        status,
        version: { increment: 1 },
        updatedAt: new Date(),
      },
    });
    if (updated.count !== 1) {
      throw DomainError.conflict(`Inbox item ${itemId} has changed`);
    }
    const item = await this.findByIdentity(itemId);
    if (!item) {
      throw DomainError.internal(`Inbox item ${itemId} was not persisted`);
    }
    return item;
  }

  async listRevisions(
    itemId: string,
    page: number,
    pageSize: number,
  ): Promise<[InboxRevision[], number]> {
    validatePage(page, pageSize);
    await this.requireOwnedItem(itemId);
    const where: Prisma.InboxRevisionWhereInput = { inboxItemId: itemId };
    const [rows, total] = await Promise.all([
      this.store.inboxRevision.findMany({
        where,
        orderBy: { revisionNumber: 'desc' },
        skip: (page - 1) * pageSize,
        take: pageSize,
      }),
      this.store.inboxRevision.count({ where }),
    ]);
    return [rows.map(assembleInboxRevision), total];
  }

  async findRevision(
    itemId: string,
    revisionId: string,
  ): Promise<InboxRevision | null> {
    const row = await this.store.inboxRevision.findFirst({
      where: {
        id: revisionId,
        inboxItemId: itemId,
        item: { workspaceId: this.workspaceId },
      },
    });
    return row ? assembleInboxRevision(row) : null;
  }

  private async captureResult(
    itemId: string,
    revisionId: string,
    revisionCreated: boolean,
  ): Promise<CapturedInboxItem> {
    const [item, revision] = await Promise.all([
      this.findByIdentity(itemId),
      this.findRevision(itemId, revisionId),
    ]);
    if (!item || !revision) {
      throw DomainError.internal(`Inbox item ${itemId} was not persisted`);
    }
    return { item, revision, revisionCreated };
  }

  private async requireOwnedItem(itemId: string): Promise<InboxItem> {
    const item = await this.findByIdentity(itemId);
    if (!item) {
      throw DomainError.notFound(`Inbox item ${itemId} not found`);
    }
    return item;
  }

  private async transaction<T>(
    operation: (store: PrismaStore) => Promise<T>,
  ): Promise<T> {
    if ('$transaction' in this.store) {
      return this.store.$transaction((transaction) => operation(transaction));
    }
    return operation(this.store);
  }
}

function listWhere(
  workspaceId: string,
  query: InboxListQuery,
): Prisma.InboxItemWhereInput {
  const search = query.query?.trim();
  return {
    workspaceId,
    ...(query.status ? { status: query.status } : {}),
    ...(query.sourceKind ? { sourceKind: query.sourceKind.trim() } : {}),
    ...(search
      ? {
          OR: [
            { title: { contains: search, mode: 'insensitive' } },
            { externalKey: { contains: search, mode: 'insensitive' } },
          ],
        }
      : {}),
  };
}

async function requireItem(
  store: PrismaStore,
  workspaceId: string,
  itemId: string,
): Promise<InboxItemRow> {
  const item = await store.inboxItem.findFirst({
    where: { id: itemId, workspaceId },
    include: ITEM_INCLUDE,
  });
  if (!item) {
    throw DomainError.notFound(`Inbox item ${itemId} not found`);
  }
  return item;
}

function requireLatestRevision(
  item: InboxItemRow,
): NonNullable<InboxItemRow['latestRevision']> {
  if (!item.latestRevision) {
    throw DomainError.internal(`Inbox item ${item.id} has no latest revision`);
  }
  return item.latestRevision;
}

function assertSameSource(
  item: InboxItemRow,
  source: NormalizedInboxSource,
): void {
  if (
    item.sourceKind !== source.sourceKind ||
    item.externalKey !== source.externalKey
  ) {
    throw DomainError.validation(
      `Inbox revision source must match item ${item.id}`,
    );
  }
}

function revisionData(
  id: string,
  inboxItemId: string,
  revisionNumber: number,
  source: NormalizedInboxSource,
  contentSha256: string,
  capturedAt: Date,
): Prisma.InboxRevisionUncheckedCreateInput {
  return {
    id,
    inboxItemId,
    revisionNumber,
    title: source.title,
    body: source.body,
    contentType: source.contentType,
    uri: source.uri,
    providerMetadata: inputJson(source.providerMetadata),
    sourceUpdatedAt: source.sourceUpdatedAt
      ? new Date(source.sourceUpdatedAt)
      : null,
    capturedAt,
    contentSha256,
  };
}

function assembleInboxItem(row: InboxItemRow): InboxItem {
  if (!row.latestRevisionId || !row.latestRevision) {
    throw DomainError.internal(`Inbox item ${row.id} has no latest revision`);
  }
  return new InboxItem(row.id, {
    workspace: new Ref(row.workspaceId),
    sourceKind: row.sourceKind,
    externalKey: row.externalKey,
    title: row.title,
    status: row.status as InboxItemStatus,
    latestRevisionId: row.latestRevisionId,
    latestRevisionSha256: row.latestRevision.contentSha256,
    revisionCount: row._count.revisions,
    version: row.version,
    createdAt: row.createdAt.toISOString(),
    updatedAt: row.updatedAt.toISOString(),
  });
}

function assembleInboxRevision(row: InboxRevisionRow): InboxRevision {
  return new InboxRevision(row.id, {
    item: new Ref(row.inboxItemId),
    revisionNumber: row.revisionNumber,
    title: row.title,
    body: row.body,
    contentType: row.contentType as InboxContentType,
    uri: row.uri,
    providerMetadata: jsonRecord(row.providerMetadata),
    sourceUpdatedAt: row.sourceUpdatedAt?.toISOString() ?? null,
    capturedAt: row.capturedAt.toISOString(),
    contentSha256: row.contentSha256,
  });
}

function jsonRecord(value: Prisma.JsonValue): Record<string, JsonValue> {
  if (!value || typeof value !== 'object' || Array.isArray(value)) {
    return {};
  }
  return value as Record<string, JsonValue>;
}

function validatePage(page: number, pageSize: number): void {
  if (
    !Number.isSafeInteger(page) ||
    page <= 0 ||
    !Number.isSafeInteger(pageSize) ||
    pageSize <= 0
  ) {
    throw DomainError.validation('page and pageSize must be greater than 0');
  }
}

File: libs/server-java/persistent/src/main/java/reengineering/ddd/evidence/persistent/associations/WorkspaceInbox.java
package reengineering.ddd.evidence.persistent.associations;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import io.github.jayclock.smartdomain.core.Ref;
import io.github.jayclock.smartdomain.mybatis.AssociationMapping;
import io.github.jayclock.smartdomain.mybatis.database.EntityList;
import jakarta.inject.Inject;
import java.time.Clock;
import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.UUID;
import reengineering.ddd.evidence.domain.CanonicalJson;
import reengineering.ddd.evidence.domain.DomainException;
import reengineering.ddd.evidence.domain.description.InboxItemDescription;
import reengineering.ddd.evidence.domain.description.InboxRevisionDescription;
import reengineering.ddd.evidence.domain.model.Inbox;
import reengineering.ddd.evidence.domain.model.InboxExtraction;
import reengineering.ddd.evidence.domain.model.InboxItem;
import reengineering.ddd.evidence.domain.model.InboxRevision;
import reengineering.ddd.evidence.domain.model.InboxStoryCandidate;
import reengineering.ddd.evidence.domain.model.InboxWorkflow;
import reengineering.ddd.evidence.domain.model.Iteration;
import reengineering.ddd.evidence.domain.model.Workspace;
import reengineering.ddd.evidence.persistent.mappers.InboxMapper;
import reengineering.ddd.evidence.persistent.mappers.InboxRows;

@AssociationMapping(entity = Workspace.class, field = "inbox", parentIdField = "workspaceId")
public final class WorkspaceInbox extends EntityList<String, InboxItem>
    implements Workspace.InboxAssociation {
  private static final TypeReference<Map<String, Object>> JSON_OBJECT = new TypeReference<>() {};

  private String workspaceId;
  @Inject private InboxMapper mapper;
  @Inject private InboxWorkflowStore workflow;
  @Inject private ObjectMapper objectMapper;
  @Inject private Clock clock;

  @Override
  protected List<InboxItem> findEntities(int from, int to) {
    return mapper.findItems(workspaceId, null, null, null, from, Math.max(to - from, 0)).stream()
        .map(this::item)
        .toList();
  }

  @Override
  protected InboxItem findEntity(String id) {
    InboxRows.ItemRow row = mapper.findItem(workspaceId, id);
    return row == null ? null : item(row);
  }

  @Override
  public int size() {
    return mapper.countItems(workspaceId, null, null, null);
  }

  @Override
  public Inbox.Page<InboxItem> list(Inbox.ListQuery query) {
    Inbox.validatePage(query.page(), query.pageSize());
    String sourceKind = optionalQuery(query.sourceKind());
    String search = optionalQuery(query.query());
    String status = query.status() == null ? null : query.status().wireValue();
    return new Inbox.Page<>(
        mapper
            .findItems(
                workspaceId,
                status,
                sourceKind,
                search,
                (query.page() - 1) * query.pageSize(),
                query.pageSize())
            .stream()
            .map(this::item)
            .toList(),
        mapper.countItems(workspaceId, status, sourceKind, search));
  }

  @Override
  public Inbox.Captured capture(Inbox.SourceInput sourceInput) {
    Inbox.HashedSource hashed = Inbox.normalizeAndHash(sourceInput);
    Inbox.Source source = hashed.source();
    mapper.lockSource(workspaceId, source.sourceKind(), source.externalKey());
    InboxRows.ItemRow existing =
        mapper.findSourceItem(workspaceId, source.sourceKind(), source.externalKey());
    if (existing != null) return appendRevision(existing.id(), sourceInput, null);

    String itemId = UUID.randomUUID().toString();
    String revisionId = UUID.randomUUID().toString();
    Instant timestamp = timestamp();
    mapper.insertItem(itemId, workspaceId, source, timestamp);
    mapper.insertRevision(
        revisionId,
        itemId,
        1,
        source,
        CanonicalJson.stringify(source.providerMetadata()),
        hashed.contentSha256(),
        timestamp);
    if (mapper.setInitialLatestRevision(workspaceId, itemId, revisionId) != 1) {
      throw DomainException.internal("Inbox item " + itemId + " was not persisted");
    }
    return captureResult(itemId, revisionId, true);
  }

  @Override
  public Inbox.Captured appendRevision(
      String itemId, Inbox.SourceInput sourceInput, String expectedLatestRevisionSha256) {
    Inbox.HashedSource hashed = Inbox.normalizeAndHash(sourceInput);
    Inbox.Source source = hashed.source();
    InboxRows.ItemRow current = mapper.lockItem(workspaceId, itemId);
    if (current == null) throw DomainException.notFound("Inbox item " + itemId + " not found");
    if (!current.sourceKind().equals(source.sourceKind())
        || !current.externalKey().equals(source.externalKey())) {
      throw DomainException.validation("Inbox revision source must match item " + itemId);
    }
    if (expectedLatestRevisionSha256 != null
        && !current.latestRevisionSha256().equals(expectedLatestRevisionSha256)) {
      throw DomainException.conflict("Inbox item " + itemId + " latest revision has changed");
    }
    if (current.latestRevisionSha256().equals(hashed.contentSha256())) {
      return captureResult(itemId, current.latestRevisionId(), false);
    }

    Instant timestamp = timestamp();
    InboxRows.RevisionRow historical = mapper.findRevisionByHash(itemId, hashed.contentSha256());
    if (historical != null) {
      updateLatest(current, source.title(), historical.id(), timestamp);
      return captureResult(itemId, historical.id(), false);
    }

    String revisionId = UUID.randomUUID().toString();
    mapper.insertRevision(
        revisionId,
        itemId,
        current.revisionCount() + 1,
        source,
        CanonicalJson.stringify(source.providerMetadata()),
        hashed.contentSha256(),
        timestamp);
    updateLatest(current, source.title(), revisionId, timestamp);
    return captureResult(itemId, revisionId, true);
  }

  @Override
  public InboxItem changeStatus(String itemId, Inbox.ItemStatus status, int expectedVersion) {
    Inbox.requireVersion(expectedVersion);
    requireItem(itemId);
    if (mapper.updateItemStatus(
            workspaceId, itemId, expectedVersion, status.wireValue(), timestamp())
        != 1) {
      throw DomainException.conflict("Inbox item " + itemId + " has changed");
    }
    return requireItem(itemId);
  }

  @Override
  public Inbox.Page<InboxRevision> listRevisions(String itemId, int page, int pageSize) {
    Inbox.validatePage(page, pageSize);
    requireItem(itemId);
    return new Inbox.Page<>(
        mapper.findRevisions(itemId, (page - 1) * pageSize, pageSize).stream()
            .map(this::revision)
            .toList(),
        mapper.countRevisions(itemId));
  }

  @Override
  public Optional<InboxRevision> findRevision(String itemId, String revisionId) {
    InboxRows.RevisionRow row = mapper.findRevision(workspaceId, itemId, revisionId);
    return Optional.ofNullable(row).map(this::revision);
  }

  @Override
  public InboxExtraction createExtraction(List<String> inboxItemIds, String requestedByUserId) {
    return workflow.createExtraction(workspaceId, inboxItemIds, requestedByUserId);
  }

  @Override
  public Optional<InboxExtraction> findExtraction(String extractionId) {
    return workflow.findExtraction(workspaceId, extractionId);
  }

  @Override
  public InboxWorkflow.ProposedCandidates proposeCandidates(
      String extractionId, int expectedVersion, List<InboxWorkflow.CandidateInput> candidates) {
    return workflow.proposeCandidates(workspaceId, extractionId, expectedVersion, candidates);
  }

  @Override
  public InboxWorkflow.CandidatePage listCandidates(InboxWorkflow.CandidateListQuery query) {
    return workflow.listCandidates(workspaceId, query);
  }

  @Override
  public Optional<InboxStoryCandidate> findCandidate(String candidateId) {
    return workflow.findCandidate(workspaceId, candidateId);
  }

  @Override
  public InboxWorkflow.CandidateDecision decideCandidate(
      String candidateId,
      String candidateSha256,
      InboxWorkflow.DecisionAction action,
      String reason,
      String decidedByUserId) {
    return workflow.decideCandidate(
        workspaceId, candidateId, candidateSha256, action, reason, decidedByUserId);
  }

  @Override
  public Iteration selectCandidate(
      InboxWorkflow.SelectCandidateInput input, String selectedByUserId) {
    return workflow.selectCandidate(workspaceId, input, selectedByUserId);
  }

  private void updateLatest(
      InboxRows.ItemRow current, String title, String revisionId, Instant timestamp) {
    if (mapper.updateLatestRevision(
            workspaceId, current.id(), current.version(), title, revisionId, timestamp)
        != 1) {
      throw DomainException.conflict("Inbox item " + current.id() + " has changed");
    }
  }

  private Inbox.Captured captureResult(String itemId, String revisionId, boolean revisionCreated) {
    InboxItem item = requireItem(itemId);
    InboxRevision revision =
        findRevision(itemId, revisionId)
            .orElseThrow(
                () -> DomainException.internal("Inbox item " + itemId + " was not persisted"));
    return new Inbox.Captured(item, revision, revisionCreated);
  }

  private InboxItem requireItem(String itemId) {
    return findByIdentity(itemId)
        .orElseThrow(() -> DomainException.notFound("Inbox item " + itemId + " not found"));
  }

  private InboxItem item(InboxRows.ItemRow row) {
    if (row.latestRevisionId() == null || row.latestRevisionSha256() == null) {
      throw DomainException.internal("Inbox item " + row.id() + " has no latest revision");
    }
    return new InboxItem(
        row.id(),
        new InboxItemDescription(
            new Ref<>(row.workspaceId()),
            row.sourceKind(),
            row.externalKey(),
            row.title(),
            Inbox.ItemStatus.parse(row.status()),
            row.latestRevisionId(),
            row.latestRevisionSha256(),
            row.revisionCount(),
            row.version(),
            row.createdAt(),
            row.updatedAt()));
  }

  private InboxRevision revision(InboxRows.RevisionRow row) {
    return new InboxRevision(
        row.id(),
        new InboxRevisionDescription(
            new Ref<>(row.inboxItemId()),
            row.revisionNumber(),
            row.title(),
            row.body(),
            Inbox.ContentType.parse(row.contentType()),
            row.uri(),
            metadata(row.providerMetadata()),
            row.sourceUpdatedAt(),
            row.capturedAt(),
            row.contentSha256()));
  }

  private Map<String, Object> metadata(String json) {
    try {
      return CanonicalJson.normalizeObject(
          objectMapper.readValue(json, JSON_OBJECT), "provider metadata");
    } catch (JsonProcessingException error) {
      throw DomainException.internal("Inbox provider metadata could not be read");
    }
  }

  private Instant timestamp() {
    return clock.instant().truncatedTo(ChronoUnit.MILLIS);
  }

  private static String optionalQuery(String value) {
    if (value == null || value.trim().isEmpty()) return null;
    return value.trim();
  }
}
